scheduler/src/cook/scheduler/constraints.clj

```diff
@@ -99,33 +90,48 @@
   (let [previous-hosts (job->previous-hosts-to-avoid job)]
     (->novel-host-constraint job previous-hosts)))
 
-(defrecord gpu-host-constraint [job needs-gpus?]
+(defn gpu-models-config-for-pool
+  "Given a pool name, obtain the valid-gpu-models entry on config for that pool"
+  [valid-gpu-models effective-pool-name]
+  (->> valid-gpu-models
+       (filter (fn [{:keys [pool-regex]}] (re-find (re-pattern pool-regex) effective-pool-name)))
+       first))
+
+(defrecord gpu-host-constraint [job]
   JobConstraint
   (job-constraint-name [this] (get-class-name this))
   (job-constraint-evaluate
     [this _ vm-attributes]
     (job-constraint-evaluate this nil vm-attributes []))
   (job-constraint-evaluate
-    [this _ vm-attributes target-vm-tasks-assigned]
-    (let [; Look at attribute and running jobs to determine if vm has gpus
-          vm-has-gpus? (or (get vm-attributes "COOK_GPU?") ; Set when putting attributes in cache
-                           (some (fn gpu-task? [{:keys [needs-gpus?]}]
-                                   needs-gpus?)
-                                 target-vm-tasks-assigned))
-          job (:job this)
-          passes? (or (and needs-gpus? vm-has-gpus?)
-                      (and (not needs-gpus?) (not vm-has-gpus?)))]
-      [passes? (when-not passes? (if (and needs-gpus? (not vm-has-gpus?))
-                                   "Job needs gpus, host does not have gpus."
-                                   "Job does not need gpus, host has gpus."))])))
+    [{:keys [job]} _ vm-attributes _]
+    (let [gpu-count-requested (-> job util/job-ent->resources :gpus (or 0))
+          gpu-model-requested (when (pos? gpu-count-requested)
+                                (or (-> job util/job-ent->env (get "COOK_GPU_MODEL"))
+                                  (:default-model (gpu-models-config-for-pool (config/valid-gpu-models) (util/job->pool-name job)))))
+          gpu-model->count-available (get vm-attributes "gpus")
+          k8s-vm? (= (get vm-attributes "compute-cluster-type") "kubernetes")
+          ; k8s VM that supports gpu models but does not have any available gpus will have a gpus map of {"model A" 0 "model B" 0 ...}
+          ; k8s VM that does not support gpu models will have an empty gpus map of {}
+
+          ; requesting gpus is only being supported for kubernetes-bound jobs
+          passes? (if k8s-vm?
```

Comment:
renamed `passes?` to `vm-satisfies-constraint?`

restructured the test to the following structure:
1. kubernetes vm
- 1a. if job requested gpus, require that the VM has enough gpus available in the same model as the job requested.
- 1b. if job did not request gpus, require that the VM does not support gpus.
2. mesos vm
- require that the job requested 0 gpus (since we are not allowing mesos jobs to request gpus).

Reply:
added some additional comments and I don't feel that we need intermediate variables now.

Reply:
Looks good.